common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -403,38 +394,78 @@ public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) throws IOExc
+      msg.shuffleId, msg.appId, msg.attemptId);
     return mergeStatuses;
   }
 
   @Override
   public void registerExecutor(String appId, ExecutorShuffleInfo executorInfo) {
     if (logger.isDebugEnabled()) {
       logger.debug("register executor with RemoteBlockPushResolver {} local-dirs {} "
-        + "num sub-dirs {}", appId, Arrays.toString(executorInfo.localDirs),
-          executorInfo.subDirsPerLocalDir);
+        + "num sub-dirs {} shuffleManager {}", appId, Arrays.toString(executorInfo.localDirs),
+        executorInfo.subDirsPerLocalDir, executorInfo.shuffleManager);
+    }
+    String shuffleManagerMeta = executorInfo.shuffleManager;
+    if (shuffleManagerMeta.contains(":")) {
+      String mergeDirInfo = shuffleManagerMeta.substring(shuffleManagerMeta.indexOf(":") + 1);
+      try {
+        ObjectMapper mapper = new ObjectMapper();
+        MergeDirectoryMeta mergeDirectoryMeta =
+          mapper.readValue(mergeDirInfo, MergeDirectoryMeta.class);
+        if (mergeDirectoryMeta.attemptId == ATTEMPT_ID_UNDEFINED) {
+          // When attemptId is -1, there is no attemptId stored in the ExecutorShuffleInfo.
+          // Only the first ExecutorRegister message can register the merge dirs
+          appsShuffleInfo.computeIfAbsent(appId, id ->
+            new AppShuffleInfo(
+              appId, mergeDirectoryMeta.attemptId,
+              new AppPathsInfo(appId, executorInfo.localDirs,
+                mergeDirectoryMeta.mergeDir, executorInfo.subDirsPerLocalDir)
+            ));
+        } else {
+          // If attemptId is not -1, there is attemptId stored in the ExecutorShuffleInfo.
+          // The first ExecutorRegister message from the same application attempt wil register
+          // the merge dirs in External Shuffle Service. Any later ExecutorRegister message
+          // from the same application attempt will not override the merge dirs. But it can
+          // be overridden by ExecutorRegister message from newer application attempt,
+          // and former attempts' shuffle partitions information will also be cleaned up.
+          ConcurrentMap<Integer, AppShuffleInfo> appShuffleInfoToBeCleanedUp =
+            Maps.newConcurrentMap();
+          appsShuffleInfo.compute(appId, (id, appShuffleInfo) -> {
+            if (appShuffleInfo == null || (appShuffleInfo != null
+              && mergeDirectoryMeta.attemptId > appShuffleInfo.attemptId)) {
+              appShuffleInfoToBeCleanedUp.putIfAbsent(appShuffleInfo.attemptId, appShuffleInfo);
+              appShuffleInfo =
+                new AppShuffleInfo(
+                  appId, mergeDirectoryMeta.attemptId,
+                  new AppPathsInfo(appId, executorInfo.localDirs,
+                    mergeDirectoryMeta.mergeDir, executorInfo.subDirsPerLocalDir));
+            }
+            return appShuffleInfo;
+          });
+          for (AppShuffleInfo appShuffleInfo: appShuffleInfoToBeCleanedUp.values()) {
```

Comment:
I wonder when there would be multiples app infos to be cleaned up?

Reply:
The semantics of using `ConcurrentMap.compute*` api's means that in theory we can have multiple registration events for different attempts simultaneously (note - this is at ESS side, so the intermediate registeration attempt by an executor could be for an application which  might have already been killed by yarn).

Having said that, @zhouyejoe is changing the Map's in this class to explicitly be `ConcurrentHashMap`, which prevents this issue from occurring due to its stricter semantics for `compute` and similar methods.

Reply:
Ok, I see the possibility of multiple registrations. But, shouldn't `appShuffleInfoToBeCleanedUp` be global rather than local?

Reply:
The change that @zhouyejoe will make to `ConcurrentHashMap` will simplify this analysis.
For `ConcurrentMap`, the `remappingFunction` passed to `compute` can get invoked multiple times - while for `ConcurrentHashMap` it is more restrictive and can be invoked only once for a key.

Given the current state (of `ConcurrentMap` instead of `ConcurrentHashMap` for `appsShuffleInfo`), I will try to elaborate the cases we have:

* (a) There is no value for `appId` in `appsShuffleInfo` - and compute ends up introducing a new value.
  * This is the simplest case - essentially ESS is finding out about a new application/attempt. In this case `appShuffleInfoToBeCleanedUp` will be empty.
  
* (b) There is an existing entry in `appsShuffleInfo` for the `appId` for the same `attemptId`.
  * In this case, `appShuffleInfoToBeCleanedUp` will be empty - and existing `appShuffleInfo` remains in `appsShuffleInfo`.
  
* (c) Variant of above - where executor registration is for an older attempt id (due to some race in registrations/allocations).
  * In this case also, `appShuffleInfoToBeCleanedUp` will be empty - and existing `appShuffleInfo` remains in `appsShuffleInfo` (since it is for newer attempt).
  
* (d) Variant of above - where executor registration is for a newer attempt id.
  * In this case, `appShuffleInfoToBeCleanedUp` will have an entry for the older attempt id - and new entry will be added to `appsShuffleInfo` (since it is for newer attempt).


In case of `ConcurrentHashMap`, all variants of the above in any order or interleaving and whatever concurrency will result in need for a single AtomicReference as was done in initial version of PR [here](https://github.com/apache/spark/blob/65fca881bc8f2184e2e01b836ab2eaeee222e44f/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java#L515)

This is because `remappingFunction` wont be called multiple times for an update, but only once.

The issue we have with `ConcurrentMap` is that as `remappingFunction` can get invoked multiple times (in case map detects concurrent interleaving updates), we can have cases where multiple (d)'s can interleave with (c).
For example:
(existing value 0) with registeration attempts 1, 2, 3 interleaving (and applied in that order) will result in `appShuffleInfoToBeCleanedUp` having entries for 0, 1, 2.
On other hand, for same case if it was applied in order of 3, 2, 1 - we will only have entry for 0 in `appShuffleInfoToBeCleanedUp`.

Not sure if this makes sense - this became a much longer explaination than I originally intended !
Hopefully, with the pending `ConcurrentHashMap` change, analysis will be simpler :-) Bottomline is, be very case when using `ConcurrentMap` instead of `ConcurrentHashMap` as the semantics of the compute* api's are not as strict.

Reply:
Thanks for the detailed explanation. Will update accordingly.

Reply:
@mridulm  Thanks for the detailed explanation. Although we're going to use `ConcurrentHashMap`, I want to make sure I understand the issue of `ConcurrentMap` clearly. But please go ahead with `ConcurrentHashMap` without blocking by my question. @zhouyejoe 

> (existing value 0) with registration attempts 1, 2, 3 interleaving (and applied in that order) will result in appShuffleInfoToBeCleanedUp having entries for 0, 1, 2.

I think `remappingFunction` here is also a local variable indeed as well as the `appShuffleInfoToBeCleanedUp`. Thus, I really can't get to know why the `appShuffleInfoToBeCleanedUp` could have multiple values at the end. With my understanding, even `remappingFunction` could be called multiple times for an update, there should be three different `remappingFunction`/`appShuffleInfoToBeCleanedUp` instances separately. So each `appShuffleInfoToBeCleanedUp` should only contains one value.

Sorry if this bothers you. I probably missed some important points. @mridulm @zhouyejoe

Reply:
It is related to the description of the [compute](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentMap.html#compute-K-java.util.function.BiFunction-) method for ConcurrentMap. ConcurrentMap is an interface, and if we take a look at how the [compute](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentSkipListMap.html#compute-K-java.util.function.BiFunction-) in ConcurrentSkipListMap is implemented, we can find that it is implemented with a for(;;) loop, where it finally calls UNSAFE.compareAndSwapObject(this, valueOffset, cmp, val). When updating the original value(origin_value) with the new value generated from the remapping function, it need to check if the current value in the (key, value) is the same with the origin_value recorded when the compute operation starts. If not, the for loop will start the next iteration and try updating the value again. In the case mentioned by Mridul above, (existing value 0) with registration attempts 1, 2, 3 interleaving, for thread which processed attempt 3, the compute can be triggered three times, where it will cleanup the former attempts 0, 1, 2. But this will also show that attempt 1 and 2 has been successfully registered in their own compute for loop. I think we can still use a single AppShuffleInfo here to store the one to be cleaned up. If we use a single AppShuffleInfo to record the one to be cleaned up, when attemp 1 gets registered, it will cleanup the stuff for attempt 0. And when attempt 2 gets registered, it will cleanup the stuff for attempt 1. There won't be cases that either attempt 1 or attempt 0 cleanup will be missed. If attempt 1 fails to get registered because attempt 2 updates the value attempt 0 first, there is no need to cleanup the stuff for attempt 1.

Reply:
I updated with ConcurrentHashMap anyway as the atomic and locking is required for our logic. Even though the current implementation of Maps.newConcurrentMap actually returns a ConcurrentHashMap, it is still better not to rely on third-party library to create the ConcurrentHashMap.